### PR TITLE
Make `UsageReporter` reusable

### PR DESCRIPTION
**Pull request by gzdunek**
The aim of this PR is to use `UsageReporter` in Teleport and Connect reporting. The main benefit of this is that we can reuse the batching/queuing logic. The main difference is anonymization: in Teleport we have only one anonymizer per reporter, while in Connect each event can be anonymized with a different anonymizer (or not anonymized at all).

What I did:
`local/usagereporter` is now generic, so it can handle any event type. It doesn't know anything about anonymization. I moved it to `lib/usagereporter` (is it a good place?).
For Teleport I created `TeleportUsageReporter` that works mostly as `UsageReporter` before. Event anonymization is now performed here. 
Connect uses `UsageReporter` directly. [Here is the PR](https://github.com/gravitational/teleport/pull/19564).

I'd like to get some feedback before I spend too much time polishing things :)

[Teleport.e] https://github.com/gravitational/teleport.e/pull/676

### CODE REVIEW

File: lib/services/usagereporter.go [hunk 1]
```diff
@@ -253,3 +294,133 @@ func ConvertUsageEvent(event *usageevents.UsageEventOneOf, identityUsername stri
...
+	usageReporter *usagereporter.UsageReporter[prehogapi.SubmitEventRequest]
+	// anonymizer is the anonymizer used for filtered audit events.
+	anonymizer utils.Anonymizer
+	// clusterName is the cluster's name, used for anonymization and as an event
+	// field.
+	clusterName types.ClusterName
+	clock       clockwork.Clock
+}
+
+func (tur *TeleportUsageReporter) SubmitAnonymizedUsageEvents(events ...UsageAnonymizable) error {
```
Comment (zmb3) [inline lib/services/usagereporter.go:309]: Since the input events are not yet anonymized, I'd consider renaming this. Maybe `AnonymizeAndSubmit(events...)`

**Review by zmb3 — COMMENTED** [review]
The generics make things a little bit hard to follow, but I don't have a better suggestion right now. I'll make another pass tomorrow and see if I have any new ideas.

File: lib/usagereporter/usagereporter.go [hunk 2]
```diff
@@ -112,40 +71,27 @@ var (
...
-
-// NewDiscardUsageReporter creates a new usage reporter that drops all events.
-func NewDiscardUsageReporter() *DiscardUsageReporter {
-	return &DiscardUsageReporter{}
-}
-
-// submitFunc is a func that submits a batch of usage events.
-type UsageSubmitFunc func(reporter *UsageReporter, batch []*SubmittedEvent) ([]*SubmittedEvent, error)
+// UsageSubmitFunc is a func that submits a batch of usage events.
+type UsageSubmitFunc[T any] func(reporter *UsageReporter[T], batch []*SubmittedEvent[T]) ([]*SubmittedEvent[T], error)
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply lib/usagereporter/usagereporter.go:82]: Good idea, I will change it

File: lib/services/usagereporter.go (hunk 1 above)
Comment (gzdunek) [reply lib/services/usagereporter.go:309]: Agree, this name was slightly confusing for me in the original implementation, but I didn't change it. I will do it.

**Comment by gzdunek** [comment]
>The generics make things a little bit hard to follow 

Yeah, lines like this are quite hard to read :/
```go
return func(reporter *usagereporter.UsageReporter[prehogapi.SubmitEventRequest], events []*usagereporter.SubmittedEvent[prehogapi.SubmitEventRequest]) ([]*usagereporter.SubmittedEvent[prehogapi.SubmitEventRequest], error) {
```

To make things a little simpler (in `tshd` part) I think I can remove `ConnectUsageReporter` struct and use `UsageReporter` directly in `tshd`. I will move adding timestamps to the electron code, so here we will receive events with all required properties. The only task of `tshd` will be to anonymize events and convert them to the `prehog` model (and then put in the queue ofc).

**Comment by timothyb89** [comment]
> > The generics make things a little bit hard to follow
> 
> Yeah, lines like this are quite hard to read :/

On the other hand it does meaningfully simplify the test code since we don't need a giant `switch` to check particular events.

**Comment by gzdunek** [comment]
I added conversion of all Connect events to unblock PR for Electron.

**Review by espadolini — APPROVED** [review]
I'll defer to @timothyb89's opinion as he's the original author, but I like this approach.

File: lib/services/usagereporter.go [hunk 3]
```diff
@@ -253,3 +298,135 @@ func ConvertUsageEvent(event *usageevents.UsageEventOneOf, identityUsername stri
...
+	usageReporter *usagereporter.UsageReporter[prehogapi.SubmitEventRequest]
+	// anonymizer is the anonymizer used for filtered audit events.
+	anonymizer utils.Anonymizer
+	// clusterName is the cluster's name, used for anonymization and as an event
+	// field.
+	clusterName types.ClusterName
+	clock       clockwork.Clock
+}
+
+func (tur *TeleportUsageReporter) AnonymizeAndSubmit(events ...UsageAnonymizable) error {
```
Comment (zmb3) [inline lib/services/usagereporter.go:313]:
```suggestion
func (t *TeleportUsageReporter) AnonymizeAndSubmit(events ...UsageAnonymizable) error {
```

File: lib/services/usagereporter.go [hunk 4]
```diff
@@ -253,3 +298,135 @@ func ConvertUsageEvent(event *usageevents.UsageEventOneOf, identityUsername stri
 		return nil, trace.BadParameter("invalid usage event type %T", event.GetEvent())
 	}
 }
+
+type TeleportUsageReporter struct {
```
Comment (zmb3) [inline lib/services/usagereporter.go:302]:
```suggestion
// TeleportUsageReporter submits Teleport usage events
// anonymized with the cluster name.
type TeleportUsageReporter struct {
```

**Review by zmb3 — APPROVED** [review]

File: lib/services/usagereporter.go (hunk 3 above)
Comment (gzdunek) [reply lib/services/usagereporter.go]: Done

File: lib/services/usagereporter.go (hunk 4 above)
Comment (gzdunek) [reply lib/services/usagereporter.go:304]: Done
